### PR TITLE
absentee,register: send a reminder emails and texts

### PR DESCRIPTION
**Pull request by liewegas**
- Register and absentee self-print: SMS + email the next day
- Register print-and-forward: SMS only in 10 minutes, SMS + email the next day

### CODE REVIEW

File: app/absentee/tasks.py [hunk 1]
```diff
@@ -122,3 +123,28 @@ def refresh_region_links():
...
+
+    if ballot_request.action.event_set.filter(event_type=EventType.DOWNLOAD).exists():
+        return
+
+    if ballot_request.phone:
+        try:
+            n = Number.objects.get(phone=ballot_request.phone)
+            n.send_sms(
+                f"We emailed your absentee ballot request form, but you haven't downloaded and printed it yet. Here is the link: {ballot_request.result_item.download_url}"
+            )
```
[the remark this answers is not in the document]
Comment (liewegas) [reply app/absentee/tasks.py:145]:
```
    def send_sms(self, text, ignore_opt_out=False):
        if self.opt_out_time and not ignore_opt_out:
            return
```

File: app/common/models.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        self.save()
+
+        try:
+            mname, fname = self.task_name.rsplit(".", 1)
+            mhandle = importlib.import_module(mname)
+            func = getattr(mhandle, fname)
+            func.delay(*self.args)
+        except Exception:
+            pass
+
```
[the remark this answers is not in the document]
Comment (liewegas) [reply app/common/models.py:31]: I worry about making something too constrained given we've done just one of these.  For instance, these two reminders have different conditions to check before sending the reminder.  Both are conditioned on the events, but I'm not sure that will always be the case.

I put this in common bc it seemed silly to create a new app for just this and nothing else seemed like a good fit.  It's easy to move, though!

In my mind this is an analog to celery that behaves better for longer-term actions.  The downside of this approach is it is a little awkward for code to inspect what future events might be scheduled and e.g. cancel them (although certainly doable by filtering for task_name and args).

@benweissmann ?

File: app/common/models.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    class Meta:
+        ordering = ["-created_at"]
+
+    def deliver(self):
+        self.started_at = datetime.datetime.now().replace(tzinfo=datetime.timezone.utc)
+        self.save()
+
+        try:
+            mname, fname = self.task_name.rsplit(".", 1)
+            mhandle = importlib.import_module(mname)
```
Comment (benweissmann) [inline app/common/models.py:26]: You don't need to import the function; you can just enqueue the task by name like we do in the re-run option in the django admin: https://github.com/vote/turnout/blob/master/app/celery_admin/celery_admin.py#L46

File: app/common/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (benweissmann) [reply app/common/models.py:31]: Yeah, I think having this take arbitrary arguments makes sense -- like Sage said, the idea is just a general mechanism for handling celery tasks with long delays in a way that's cheap and reliable.

File: app/common/models.py (hunk 3 above)
Comment (liewegas) [reply app/common/models.py]: Ah, sooo much better, thanks!

**Comment by liewegas** [comment]
One question: I include the download link in the SMS, but most people read these on their phone and/or can't print from their phone.  Better to just say we resent the email?  Or say both?

**Comment by nickcatal** [comment]
I think both?

**Comment by liewegas** [comment]
> I think both?

updated

**Pull request merged**